web/src/reload_setup.ts

```diff
@@ -55,11 +56,47 @@ export function initialize(): void {
         // exist, but be log it so that it's available for future
         // debugging if an exception happens later.
         blueslip.info("Invalid hash change reload token");
-        message_view.changehash("", trigger);
+        message_view.changehash("", "reload");
         return;
     }
     ls.remove(hash_fragment);
 
+    const parsed = reload_metadata_schema.safeParse(fragment);
+    if (parsed.success) {
+        const data = parsed.data.reload_data;
+        if (data.compose_active_draft_id !== undefined) {
+            const draft = drafts.draft_model.getDraft(data.compose_active_draft_id);
+            if (draft === false) {
+                blueslip.warn("Tried to restore a draft that didn't exist.");
+            } else {
+                compose_actions.start({...draft, message_type: draft.type});
+                if (data.send_draft_immediately) {
+                    compose.finish();
+                }
+            }
+        }
+
+        // We only restore pointer and offset for the current narrow, even if
+        // there are narrows that were cached before the reload, they are no
+        // longer cached after the reload. We could possibly store the pointer
+        // and offset for these narrows but it might lead to a confusing
+        // experience if the user gets back to these narrow much later (maybe days)
+        // and finds them at a random position in the narrow which they didn't
+        // navigate to while they were trying to just get to the latest unread
+        // message in that narrow which will now take more effort to find.
+        message_fetch.set_initial_pointer_and_offset({
+            narrow_pointer: data.message_view_pointer,
+            narrow_offset: data.message_view_scroll_offset,
+        });
+
+        activity.set_new_user_input(false);
+        message_view.changehash(data.hash, "reload");
+    } else {
+        reload_with_legacy_data(fragment);
+    }
+}
+
+function reload_with_legacy_data(fragment: unknown): void {
```

Comment:
This name is confusing; it sounds like we're going to trigger a reload. I think we want `load_from_legacy_data` seems right. 

As a bigger refactor, I think you want the function to just convert whatever was stored there into the modern object, and then have both code paths use the modern object.

Repeating the `message_fetch.set_initial_pointer_and_offset` and similar is going to lead to bugs as this code churns, and it's pretty safe from that if we just have a single copy of the actual functionality.

Reply:
Renamed 👍 

I considered that refactor, but the modern object only has a `draft_id` and the old logic doesn't rely on a `draft_id` existing, so I think I'd have to make sure we saved a draft with that draft data, but that might overwrite existing drafts... So then probably the right call is to turn the modern data into an object with the message data in it, but that's uglier than the newer modern data code. So I just left them separate.